File: backend/src/main/java/com/drinkit/service/CartService.java
package com.drinkit.service;

import java.util.List;
import java.util.Optional;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import com.drinkit.entity.Cart;
import com.drinkit.entity.CartItem;
import com.drinkit.repository.CartItemRepository;
import com.drinkit.repository.CartRepository;

@Service
public class CartService {

    @Autowired
    private CartRepository cartRepository;

    @Autowired
    private CartItemRepository cartItemRepository;


    // Create Cart for User
    public Cart createCart(Long userId) {

        Optional<Cart> existingCart =
                cartRepository.findByUserId(userId);

        if (existingCart.isPresent()) {
            return existingCart.get();
        }

        Cart cart = new Cart();

        cart.setUserId(userId);

        return cartRepository.save(cart);
    }


    // Add Product to Cart
 // Add Product to Cart
    public CartItem addProduct(CartItem cartItem) {

        Optional<Cart> cart =
                cartRepository.findById(cartItem.getCartId());

        if (cart.isEmpty()) {
            return null;
        }

        // Check if product already exists in this cart
        List<CartItem> existingItems =
                cartItemRepository.findByCartId(cartItem.getCartId());

        for (CartItem existingItem : existingItems) {

            if (existingItem.getProductId()
                    .equals(cartItem.getProductId())) {

                existingItem.setQuantity(
                        existingItem.getQuantity()
                                + cartItem.getQuantity()
                );

                existingItem.setTotalPrice(
                        existingItem.getPrice()
                                * existingItem.getQuantity()
                );

                return cartItemRepository.save(existingItem);
            }
        }

        // New product
        cartItem.setTotalPrice(
                cartItem.getPrice()
                        * cartItem.getQuantity()
        );

        return cartItemRepository.save(cartItem);
    }


    // Get All Cart Items
    public List<CartItem> getAllCartItems() {

        return cartItemRepository.findAll();
    }


    // Get Items By Cart Id
    public List<CartItem> getCartItems(Long cartId) {

        return cartItemRepository.findByCartId(cartId);
    }


    // Update Quantity
    public CartItem updateQuantity(
            Long id,
            Integer quantity) {

        CartItem item =
                cartItemRepository
                        .findById(id)
                        .orElse(null);

        if (item == null) {
            return null;
        }

        item.setQuantity(quantity);

        item.setTotalPrice(
                item.getPrice() * quantity
        );

        return cartItemRepository.save(item);
    }


    // Delete Cart Item
    public String deleteCartItem(Long id) {

        CartItem item =
                cartItemRepository
                        .findById(id)
                        .orElse(null);

        if (item == null) {
            return "Cart Item Not Found";
        }

        cartItemRepository.delete(item);

        return "Item Removed Successfully";
    }


    // Clear Cart
    public String clearCart(Long cartId) {

        List<CartItem> items =
                cartItemRepository.findByCartId(cartId);

        cartItemRepository.deleteAll(items);

        return "Cart Cleared Successfully";
    }


    // Calculate Total Amount
    public Double calculateTotal(Long cartId) {

        List<CartItem> items =
                cartItemRepository.findByCartId(cartId);

        double total = 0;

        for (CartItem item : items) {

            total += item.getTotalPrice();

        }

        return total;
    }
}

File: frontend/src/data/cartService.js
import axios from "axios";
import { getToken } from "./authService";

const API_URL =
  `${import.meta.env.VITE_API_URL}/cart`;


const getHeaders = () => ({
  headers: {
    Authorization:
      `Bearer ${getToken()}`
  }
});


// Create / Get Cart
export const createCart = async (userId) => {

  const response =
    await axios.post(
      `${API_URL}/create/${userId}`,
      {},
      getHeaders()
    );

  return response.data;
};


// Add Product
export const addProductToCart =
  async (cartItem) => {

    const response =
      await axios.post(
        `${API_URL}/add`,
        cartItem,
        getHeaders()
      );

    return response.data;
  };


// Get Cart Items
export const getCartItems =
  async (cartId) => {

    const response =
      await axios.get(
        `${API_URL}/${cartId}`,
        getHeaders()
      );

    return response.data;
  };


// Update Quantity
export const updateCartQuantity =
  async (id, quantity) => {

    const response =
      await axios.put(
        `${API_URL}/update/${id}/${quantity}`,
        {},
        getHeaders()
      );

    return response.data;
  };


// Delete Cart Item
export const deleteCartItem =
  async (id) => {

    const response =
      await axios.delete(
        `${API_URL}/delete/${id}`,
        getHeaders()
      );

    return response.data;
  };


// Clear Cart
export const clearCart =
  async (cartId) => {

    const response =
      await axios.delete(
        `${API_URL}/clear/${cartId}`,
        getHeaders()
      );

    return response.data;
  };


// Get Total
export const getCartTotal =
  async (cartId) => {

    const response =
      await axios.get(
        `${API_URL}/total/${cartId}`,
        getHeaders()
      );

    return response.data;
  };